test/model/message_list_test.dart

```diff
@@ -0,0 +1,241 @@
+
+      final updatedMessage = messageList.messages[0];
+      check(updatedMessage).identicalTo(message);
+      check(listenersNotified).equals(true);
+
+      check(message)
+        ..content.equals(newContent)
+        ..lastEditTimestamp.equals(newTimestamp)
+        ..flags.equals(newFlags)
+        ..isMeMessage.equals(true);
+    });
+
+    test('update event is ignored when message is not in the message list', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, edited</p>";
+      const newTimestamp = 99999;
+
+      final mockMessage = eg.streamMessage(id: 243, stream: stream, content: oldContent);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: 972,
+        messageIds: [972],
+        flags: mockMessage.flags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp,
+        userId: userId,
+      );
+
+      final message = messageList.messages[0];
+      check(message).content.equals(oldContent);
+
+      var listenersNotified = false;
+
+      messageList.addListener(() { listenersNotified = true; });
+      messageList.maybeUpdateMessage(updateEvent);
+
+      check(listenersNotified).equals(false);
+      check(message).content.equals(oldContent);
+
+    });
+    test('rendering-only update does not change timestamp', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, world</p> <div>Some link preview</div>";
+      const newTimestamp = 99999;
+
+      final mockMessage = eg.streamMessage(id: 972, stream: stream, content: oldContent);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: 972,
+        messageIds: [972],
+        flags: mockMessage.flags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp,
+        renderingOnly: true,
+        userId: null,
+      );
+
+      final message = messageList.messages[0];
+      messageList.maybeUpdateMessage(updateEvent);
+      check(message)
+        ..content.equals(newContent)
+        ..lastEditTimestamp.isNull();
+    });
+
+    test('rendering-only update does not change timestamp (for old server versions)', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, world</p> <div>Some link preview</div>";
+      const newTimestamp = 99999;
+
+      final mockMessage = eg.streamMessage(id: 972, stream: stream, content: oldContent);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
```

Comment:
Ah my assumption was that for older server versions, it will omit the `renderingOnly` key entirely.

Reply:
Ah. Yeah, the server will leave it out of its response — old servers don't know about anything named `rendering_only`. But then in the `UpdateMessageEvent` object that we deserialize from the server's JSON, that class has a field `renderingOnly` with type `bool?`:
```dart
  final bool? renderingOnly; // TODO(server-5)
```
which means the object will always have some value in a spot called `renderingOnly`, and that value is of type `bool?` — which means it's either `null` or a value of type `bool`.

So when we deserialize the JSON and there's no `rendering_only` key in that JSON object, the value we use for the `renderingOnly` field in the resulting `UpdateMessageEvent` is naturally `null`.

This does mean that if a key is present in some JSON object but with a value of `null`, we usually don't distinguish that from the key being absent — either one will deserialize to a field with value `null`. This has been fine, because it's rare for an API to have both those possibilities for a given key and have them mean different things. (Certainly it's rare in the Zulip API, and I'd expect it to be rare in sane APIs generally.) There's only one such instance I can think of in the Zulip API; here's discussion on how we expect to handle it:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/flutter.3A.20omitted.20vs.2E.20null.20in.20JSON/near/1551522

Reply:
In this test, if there weren't a way to express explicitly in the code that this event came without a (non-null) `rendering_only`, I'd want to still make it explicit by using a comment — like
```dart
        // renderingOnly omitted
```
at the spot where the `renderingOnly: true,` (or whatever) would have appeared.